quantgov/estimator/estimation.py

```diff
@@ -0,0 +1,233 @@
+            for label_classes in (
+                est.classes_ for est in model.model.steps[-1][-1].estimators_
+            )
+        )
+    predicted = pipeline.predict_proba(texts)
+    for i, docidx in enumerate(streamer.index):
+        yield docidx, tuple(label_predictions[i, truecols[j]]
+                            for j, label_predictions in enumerate(predicted))
+
+
+def estimate_probability_multiclass(vectorizer, model, streamer):
+    """
+    Generate probabilities for a one-label, multiclass estimator
+
+    Arguments:
+        * vectorizer: a sklearn Vectorizer (or pipeline)
+        * model: a quantgov.estimator.Estimator
+        * streamer: a quantgov.corpora.CorpusStreamer
+
+    Yields:
+        2-tuples of docindex, probability
+
+    """
+    pipeline = get_pipeline(vectorizer, model)
+    texts = (doc.text for doc in streamer)
+    yield from zip(streamer.index, pipeline.predict_proba(texts))
+
+
+def estimate_probability_multilabel_multiclass(vectorizer, model, streamer):
+    """
+    Generate probabilities for a multilabel, multiclass estimator
+
+    Arguments:
+        * vectorizer: a sklearn Vectorizer (or pipeline)
+        * model: a quantgov.estimator.Estimator
+        * streamer: a quantgov.corpora.CorpusStreamer
+
+    Yields:
+        2-tuples of docindex, probability
+
+    """
+    pipeline = get_pipeline(vectorizer, model)
+    texts = (doc.text for doc in streamer)
+    predicted = pipeline.predict_proba(texts)
+    for i, docidx in enumerate(streamer.index):
+        yield docidx, tuple(label_predictions[i]
+                            for label_predictions in predicted)
+
+
+def is_multiclass(classes):
+    """
+    Returns True if values in classes are anything but 1, 0, True, or False,
+    otherwise returns False.
+    """
+    try:
+        return len(set(int(i) for i in classes) - {0, 1}) != 0
+    except ValueError:
+        return True
+
+
+def estimate(vectorizer, model, corpus, probability, outfile):
+    """
+    Estimate label values for documents in corpus
+
+    Arguments:
+
+        * **vectorizer**: joblib-saved vectorizer
+        * **model**: saved `quantgov.estimator.Model` object
+        * **corpus**: path to a quantgov corpus
+        * **probability**: if True, predict probability
+        * **outfile**: open file object for writing results
+    """
+    streamer = corpus.get_streamer()
+    writer = csv.writer(outfile)
+    if len(model.label_names) > 1:
+        multilabel = True
+        try:
+            multiclass = any(is_multiclass(i) for i in model.model.classes_)
+        except AttributeError:
+            multiclass = any(
+                is_multiclass(i.classes_) for i in
+                model.model.steps[-1][-1].estimators_
+            )
+    else:
+        multilabel = False
+        multiclass = is_multiclass(model.model.classes_)
+
+    # TODO: This is very ugly and complicated and should probably be refactored
```

Comment:
Would it potentially make sense to separate the writing into a separate function and yield results from the relevant option here? That wouldn't resolve the nested generators which is the largest issue in readability for me personally but might help a bit.

Reply:
Let's definitely take refactoring this as an action item. I feel like there's probably a way to detect the format of the prediction and turn that into rows automatically.